paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/PrimaryKeyLRULookupTable.java

```diff
@@ -0,0 +1,117 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.flink.lookup;
+
+import org.apache.paimon.data.BinaryRow;
+import org.apache.paimon.data.InternalRow;
+import org.apache.paimon.disk.IOManagerImpl;
+import org.apache.paimon.io.DataFileMeta;
+import org.apache.paimon.table.FileStoreTable;
+import org.apache.paimon.table.query.TableQuery;
+import org.apache.paimon.table.source.DataSplit;
+import org.apache.paimon.table.source.Split;
+import org.apache.paimon.utils.Projection;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+
+/** Lookup table for primary key which supports to read the LSM tree directly. */
+public class PrimaryKeyLRULookupTable implements LookupTable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PrimaryKeyLRULookupTable.class);
+
+    private final FixedBucketKeyExtractor extractor;
+
+    private final TableQuery tableQuery;
+
+    public PrimaryKeyLRULookupTable(FileStoreTable table, int[] projection, File path) {
+        this.tableQuery =
+                table.newTableQuery()
+                        .withValueProjection(Projection.of(projection).toNestedIndexes())
+                        .withIOManager(new IOManagerImpl(path.toString()));
+        this.extractor = new FixedBucketKeyExtractor(table.schema());
+    }
+
+    @Override
+    public List<InternalRow> get(InternalRow key) throws IOException {
+        extractor.setRecord(key);
+        int bucket = extractor.bucket();
+        BinaryRow partition = extractor.partition();
+
+        InternalRow kv = tableQuery.lookup(partition, bucket, key);
+        if (kv == null) {
+            return Collections.emptyList();
+        } else {
+            return Collections.singletonList(kv);
+        }
+    }
+
+    @Override
+    public void refresh(List<Split> splits) {
+        for (Split split : splits) {
+            if (!(split instanceof DataSplit)) {
+                throw new IllegalArgumentException("Unsupported split: " + split.getClass());
+            }
+            BinaryRow partition = ((DataSplit) split).partition();
+            int bucket = ((DataSplit) split).bucket();
+            List<DataFileMeta> before = ((DataSplit) split).beforeFiles();
+            List<DataFileMeta> after = ((DataSplit) split).dataFiles();
+
+            tableQuery.refreshFiles(partition, bucket, before, after);
+        }
+    }
+
+    @Override
+    public Predicate<InternalRow> recordFilter() {
+        throw new UnsupportedOperationException();
```

Comment:
So many `UnsupportedOperationException`s, you could add an interface `FullCacheLookupTable`?

Reply:
yes, it looks better